src/Raven.Server/Documents/Indexes/Workers/HandleReferencesBase.ReferencesState.cs

```diff
@@ -0,0 +1,110 @@
+﻿using System.Collections.Generic;
+using Raven.Server.ServerWide.Context;
+
+namespace Raven.Server.Documents.Indexes.Workers
+{
+    public abstract partial class HandleReferencesBase
+    {
+        private class ReferencesState
+        {
+            private readonly Dictionary<string, ReferenceState> _lastIdPerCollectionForDocuments = new Dictionary<string, ReferenceState>();
+
+            private readonly Dictionary<string, ReferenceState> _lastIdPerCollectionForTombstones = new Dictionary<string, ReferenceState>();
+
+            public Reference For(ActionType actionType, string collection)
+            {
+                return new Reference(GetDictionary(actionType), collection);
+            }
+
+            public InMemoryReferencesInfo GetReferencesInfo(string collection)
+            {
+                return new InMemoryReferencesInfo
+                {
+                    ParentItemEtag = new Reference(GetDictionary(ActionType.Document), collection).GetLastIndexedParentEtag(),
+                    ParentTombstoneEtag = new Reference(GetDictionary(ActionType.Tombstone), collection).GetLastIndexedParentEtag(),
+                };
+            }
+
+            public void Clear(ActionType actionType)
+            {
+                var dictionary = GetDictionary(actionType);
+                dictionary.Clear();
+            }
+
+            private Dictionary<string, ReferenceState> GetDictionary(ActionType actionType)
+            {
+                return actionType == ActionType.Document
+                    ? _lastIdPerCollectionForDocuments
+                    : _lastIdPerCollectionForTombstones;
+            }
+
+            public class ReferenceState
+            {
+                public string ReferencedItemId;
+                public string NextItemId;
+                public long ReferencedItemEtag;
+                public long LastIndexedParentEtag;
+            }
+
+            public class Reference
+            {
+                private readonly Dictionary<string, ReferenceState> _dictionary;
+                private readonly string _collection;
+
+                public Reference(Dictionary<string, ReferenceState> dictionary, string collection)
+                {
+                    _dictionary = dictionary;
+                    _collection = collection;
+                }
+
+                public void Set(HandleReferencesBase.Reference reference, string itemId, long lastIndexedParentEtag, TransactionOperationContext indexContext)
```

Comment:
This method is supposed to be called just _one_ time in a given batch because of below `AfterCommitWhenNewReadTransactionsPrevented ` action (so no guarantee in order or processing attached actions). Is that correct?

If yes, then can we add some debug assertion for that. Relaying on the assumption that this is supposed to be called just once is a bit fragile. I'm for adding something that will protect us in the future.

Reply:
it can be called multiple times during a single indexing batch, but only once per collection

Reply:
Okay. Then we can add some debug code that will validate that this is called only once per collection in a batch?

Reply:
Done